### PR TITLE
ENH: Add PYVISTA_VTK_DATA environment variable to use local vtk-data

### LINKED ISSUE #1588: Feature Request: Allow for easier generation of docs with no internet connection
**Describe the bug, what's wrong, and what you expect:**

One roadblock I have in developing the documentation is that it is required to be able to download the data using the `examples.download_*` functions.  It is easier for me to generate the documentation on a system without direct access to download these files.

I have been able to get around this somewhat by cloning the vtk-data repository and then copying over all the files under Data.  In this way I get most of the files, but certain data files are zipped or inside nested folders.  It would be convenient if there was an option to allow the examples directory to be built from a local vtk-data repo source rather than having to download it.

This _might_ speed up the CI too, especially if the repo data could be shared amongst the test jobs.

### PR DESCRIPTION
**Pull request by MatthewFlamm**
### Overview

This PR enables use of a local vtk-data repository to use for examples instead of retrieving from the internet.  This is useful when running from a machine that is not connected to the internet.

Closes #1588 

### Details

The documentation cannot still be 100% generated without an internet connection as there are a few other external data sources not related to vtk-data

This is currently not tested in the CI, but I was able to generate (most of) the documentation from a system without internet access, and run the test suite.
image

The new documentation on this feature:
image

### CODE REVIEW

**Comment by akaszynski** [comment]
> The documentation cannot still be 100% generated without an internet connection as there are a few other external data sources not related to vtk-data

Can we just copy this to the vtk-data repo?

**Comment by MatthewFlamm** [comment]
I suppose it depends on license compatibility.

The ones I noted when building the documentation were these:

https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/DamagedHelmet  <- CCA non-commercial, but it links to the model in sketchfab, and I wasn't able to confirm this.

https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/SheenChair <- CC-BY 4.0

https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/GearboxAssy <- no stated license

https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/Avocado <- CC0 1.0

https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/CesiumMilkTruck <- CC BY 4.0

We may want to remove GearBoxAssy completely from the documentation based on the lack of license.

File: doc/extras/vtk_data.rst [hunk 1]
```diff
@@ -0,0 +1,16 @@
+
+Using local vtk-data
+=====================
```
Comment (akaszynski) [inline doc/extras/vtk_data.rst:3]:
```suggestion
Using Local VTK-Data
====================
```

File: pyvista/__init__.py [hunk 2]
```diff
@@ -45,6 +46,11 @@
 except KeyError:
     pass
 
+# If available, a local vtk-data instance will be used for examples
+VTK_DATA_PATH: Optional[str] = None
+if 'PYVISTA_VTK_DATA' in os.environ:
+    VTK_DATA_PATH = os.environ['PYVISTA_VTK_DATA']
```
Comment (akaszynski) [inline pyvista/__init__.py:52]: Optionally check if this is a valid path here.  Might clear up some headaches when debugging this.

**Review by akaszynski — APPROVED** [review]
Approved with two minor suggestions.  We can consider moving the gltf files over; let's make a separate PR for the gearbox assembly since the license is unclear.

File: pyvista/__init__.py (hunk 2 above)
Comment (MatthewFlamm) [reply pyvista/__init__.py:52]: I will add a warning when it doesn't match in `pyvista.__init__`.  This seems better than not allowing pyvista to import by raising an error.

Comment (adeak) [reply pyvista/__init__.py:52]: Should we perhaps raise a loud error here instead? Someone set the envvar for a reason, and they might miss the warning (perhaps in CI?). What kind of use cases are we preparing for with this feature? If "silently" (with a warning) downloading is OK considering those use cases, then just a warning is fine as well.

Comment (MatthewFlamm) [reply pyvista/__init__.py:52]: I'm okay with raising an error here too. I don't have a strong feeling.

We could also warn on import and raise an error with an understandable error message when the user tries to download.  This seems like the best of both worlds to me.

File: pyvista/__init__.py [hunk 3]
```diff
@@ -45,6 +46,15 @@
...
+    if not os.path.isdir(VTK_DATA_PATH):
+        warnings.warn(f"VTK_DATA_PATH: {VTK_DATA_PATH} is an invalid path")
+    if not os.path.isdir(os.path.join(VTK_DATA_PATH, 'Data')):
+        warnings.warn(f"VTK_DATA_PATH: {os.path.join(VTK_DATA_PATH, 'Data')} does not exist")
```
Comment (akaszynski) [inline pyvista/__init__.py:56]: Agreed with this.  You start in a bad configuration but you should still be allowed to execute code that doesn't require examples.  Only when you actually use it should you raise errors.

File: pyvista/examples/downloads.py [hunk 4]
```diff
@@ -53,21 +48,60 @@ def _retrieve_file(url, filename):
...
-    url = _get_vtk_file_url(filename)
-    return _retrieve_file(url, filename)
+    if pyvista.VTK_DATA_PATH is None:
+        url = _get_vtk_file_url(filename)
+        retriever = partial(_http_request, url)
+    else:
+        if not os.path.isdir(pyvista.VTK_DATA_PATH):
+            raise ValueError('VTK data repository path does not exist')
+        if not os.path.isdir(os.path.join(pyvista.VTK_DATA_PATH), 'Data'):
+            raise ValueError("VTK data repository does not have 'Data' folder")
```
Comment (akaszynski) [inline pyvista/examples/downloads.py:102]:
```suggestion
            raise FileNotFoundError('VTK data repository path does not exist')
        if not os.path.isdir(os.path.join(pyvista.VTK_DATA_PATH), 'Data'):
            raise FileNotFoundError("VTK data repository does not have 'Data' folder")
```

**Comment by akaszynski** [comment]
Made a minor change, otherwise still LGTM.

**Comment by adeak** [comment]
Just to bikeshed a bit: should we add the actual paths to the error messages? Might also help debugging.

File: pyvista/examples/downloads.py [hunk 5]
```diff
@@ -53,21 +48,60 @@ def _retrieve_file(url, filename):
...
 
 def _download_file(filename):
-    url = _get_vtk_file_url(filename)
-    return _retrieve_file(url, filename)
+    if pyvista.VTK_DATA_PATH is None:
+        url = _get_vtk_file_url(filename)
+        retriever = partial(_http_request, url)
+    else:
+        if not os.path.isdir(pyvista.VTK_DATA_PATH):
+            raise FileNotFoundError('VTK data repository path does not exist')
```
Comment (akaszynski) [inline pyvista/examples/downloads.py:100]:
```suggestion
            raise FileNotFoundError(f'VTK data repository path does not exist at:\n\n{pyvista.VTK_DATA_PATH}')
```

File: pyvista/examples/downloads.py [hunk 6]
```diff
@@ -53,21 +48,60 @@ def _retrieve_file(url, filename):
...
+        if not os.path.isdir(os.path.join(pyvista.VTK_DATA_PATH), 'Data'):
+            raise FileNotFoundError("VTK data repository does not have 'Data' folder")
```
Comment (akaszynski) [inline pyvista/examples/downloads.py:102]:
```suggestion
            raise FileNotFoundError(f'VTK data repository does not have 'Data' folder at:\n\n{pyvista.VTK_DATA_PATH}')
```

**Comment by akaszynski** [comment]
> Just to bikeshed a bit: should we add the actual paths to the error messages? Might also help debugging.

Added.

File: pyvista/examples/downloads.py [hunk 7]
```diff
@@ -53,21 +48,60 @@ def _retrieve_file(url, filename):
...
-    url = _get_vtk_file_url(filename)
-    return _retrieve_file(url, filename)
+    if pyvista.VTK_DATA_PATH is None:
+        url = _get_vtk_file_url(filename)
+        retriever = partial(_http_request, url)
+    else:
+        if not os.path.isdir(pyvista.VTK_DATA_PATH):
+            raise FileNotFoundError(f'VTK data repository path does not exist at:\n\n{pyvista.VTK_DATA_PATH}')
+        if not os.path.isdir(os.path.join(pyvista.VTK_DATA_PATH), 'Data'):
+            raise FileNotFoundError(f'VTK data repository does not have 'Data' folder at:\n\n{pyvista.VTK_DATA_PATH}')
```
Comment (adeak) [inline pyvista/examples/downloads.py:102]:
```suggestion
            raise FileNotFoundError(f'VTK data repository does not have "Data" folder at:\n\n{pyvista.VTK_DATA_PATH}')
```

File: pyvista/examples/downloads.py [hunk 8]
```diff
@@ -99,7 +99,7 @@ def _download_file(filename):
         if not os.path.isdir(pyvista.VTK_DATA_PATH):
             raise FileNotFoundError(f'VTK data repository path does not exist at:\n\n{pyvista.VTK_DATA_PATH}')
         if not os.path.isdir(os.path.join(pyvista.VTK_DATA_PATH), 'Data'):
-            raise FileNotFoundError(f'VTK data repository does not have 'Data' folder at:\n\n{pyvista.VTK_DATA_PATH}')
+            raise FileNotFoundError(f'VTK data repository does not have "Data" folder at:\n\n{pyvista.VTK_DATA_PATH}')
```
Comment (akaszynski) [inline pyvista/examples/downloads.py:102]: Thanks!

**Pull request merged**